### PR TITLE
Generalized map iterator and accessor

### PR DESCRIPTION
**Pull request by grixa**
CC @dkorolev @mzhurovich

### CODE REVIEW

File: Bricks/util/iterator.h [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+  const value_t* operator->() const { return is_unique_ptr<mapped_t>::pointer(iterator_->second); }
+};
+
+template <typename MAP>
+struct GenericMapAccessor final {
+  using iterator_t = GenericMapIterator<MAP>;
+  using key_t = typename MAP::key_type;
+  const MAP& map_;
+
+  GenericMapAccessor(const MAP& map) : map_(map) {}
```
Comment (dkorolev) [inline Bricks/util/iterator.h]: `explicit`.

File: Bricks/util/test.cc [hunk 2]
```diff
@@ -517,6 +518,44 @@ TEST(Util, LazyInstantiation) {
   EXPECT_EQ("400:7", bar_y_q.InstantiateAsUniquePtrWithExtraParameter(400)->AsString());
 }
 
+TEST(Util, Iterator) {
+  typedef std::map<int, std::string> map1_t;
+  map1_t map1;
+  auto accessor1 = current::GenericMapAccessor<map1_t>(map1);
```
Comment (dkorolev) [inline Bricks/util/test.cc]: `const auto`?

File: Bricks/util/test.cc [hunk 3]
```diff
@@ -517,6 +518,44 @@ TEST(Util, LazyInstantiation) {
...
+  EXPECT_TRUE(accessor1.Empty());
+  EXPECT_EQ(0u, accessor1.Size());
+  EXPECT_EQ(accessor1.begin(), accessor1.end());
```
Comment (dkorolev) [inline Bricks/util/test.cc]: This probably best as `EXPECT_TRUE(accessor1.begin() == accessor1.end())`. What the `EXCEPT_EQ` would print as `Actual:` and `Expected:` would be unreadable anyway.

File: Bricks/util/test.cc [hunk 4]
```diff
@@ -517,6 +518,44 @@ TEST(Util, LazyInstantiation) {
...
+  for (char c = 'a'; c <= 'z'; ++c) {
+    map1[c - 'a'] = c;
+  }
+  EXPECT_FALSE(accessor1.Empty());
+  EXPECT_EQ(26u, accessor1.Size());
+  EXPECT_FALSE(accessor1.begin() == accessor1.end());
+  std::string str1;
+  for (const auto& element : accessor1) {
+    str1 += current::ToString(element);
```
Comment (dkorolev) [inline Bricks/util/test.cc]: Isn't `current::ToString()` redundant here?

File: Bricks/util/test.cc [hunk 5]
```diff
@@ -517,6 +518,44 @@ TEST(Util, LazyInstantiation) {
   EXPECT_EQ("400:7", bar_y_q.InstantiateAsUniquePtrWithExtraParameter(400)->AsString());
 }
 
+TEST(Util, Iterator) {
+  typedef std::map<int, std::string> map1_t;
```
Comment (dkorolev) [inline Bricks/util/test.cc]: `typedef` -> `using`?

File: Bricks/util/test.cc [hunk 6]
```diff
@@ -517,6 +518,44 @@ TEST(Util, LazyInstantiation) {
...
+  EXPECT_FALSE(accessor1.Empty());
+  EXPECT_EQ(26u, accessor1.Size());
+  EXPECT_FALSE(accessor1.begin() == accessor1.end());
+  std::string str1;
+  for (const auto& element : accessor1) {
+    str1 += current::ToString(element);
+  }
+  EXPECT_EQ("abcdefghijklmnopqrstuvwxyz", str1);
+
+  typedef std::map<int, std::unique_ptr<std::string>> map2_t;
```
Comment (dkorolev) [inline Bricks/util/test.cc]: `typedef` -> `using`?

File: Bricks/util/test.cc [hunk 7]
```diff
@@ -517,6 +518,44 @@ TEST(Util, LazyInstantiation) {
...
+  map2_t map2;
+  auto accessor2 = current::GenericMapAccessor<map2_t>(map2);
+  EXPECT_TRUE(accessor2.Empty());
+  EXPECT_EQ(0u, accessor2.Size());
+  EXPECT_EQ(accessor2.begin(), accessor2.end());
+  for (auto it = accessor1.begin(), eit = accessor1.end(); it != eit; ++it) {
```
Comment (dkorolev) [inline Bricks/util/test.cc]: @sompylasar you gonna love @grixa's for-loop declaration style. :-)

File: Bricks/util/test.cc [hunk 8]
```diff
@@ -517,6 +518,44 @@ TEST(Util, LazyInstantiation) {
...
+    map2[it.key()] = std::make_unique<std::string>(*it);
+  }
+  EXPECT_FALSE(accessor2.Empty());
+  EXPECT_EQ(26u, accessor2.Size());
+  EXPECT_FALSE(accessor2.begin() == accessor2.end());
+  std::string str2;
+  for (const auto& element : accessor2) {
+    str2 += current::ToString(element);
```
Comment (dkorolev) [inline Bricks/util/test.cc]: Intuitively, `current::ToString` should not be required here too, right?

File: Bricks/util/test.cc [hunk 9]
```diff
@@ -517,6 +518,44 @@ TEST(Util, LazyInstantiation) {
...
+  }
+  EXPECT_EQ("abcdefghijklmnopqrstuvwxyz", str2);
```
Comment (dkorolev) [inline Bricks/util/test.cc]: Perhaps initialize the 2nd string with `(c - 'a' + 'A')` (or the 1st one with capital and the 2nd one with lowercase), to reduce room for human error?

Very clean test BTW.

**Comment by dkorolev** [comment]
@grixa LGTM++, comments are mostly style nits. Also, perhaps `Pointer` and `Extract` capitalized, instead of `Pointer` and `Extract`?

File: Bricks/util/test.cc (hunk 7 above)
Comment (sompylasar) [inline Bricks/util/test.cc]: `for (var ic = arr.length, i = 0; i < ic; ++i)` – I still prefer to define end before begin to be able to reference it in the `i = 0` or `i = ic - 1` initializer clause.

File: Bricks/util/test.cc [hunk 10]
```diff
@@ -517,6 +518,44 @@ TEST(Util, LazyInstantiation) {
   EXPECT_EQ("400:7", bar_y_q.InstantiateAsUniquePtrWithExtraParameter(400)->AsString());
 }
 
+TEST(Util, Iterator) {
```
Comment (mzhurovich) [inline Bricks/util/test.cc]: `Iterator`->`GenericMapAccessor`?

**Comment by mzhurovich** [comment]
@grixa LGTM and ready to merge!
Minor nit: let's rename the test to be more specific on what's going on there.

File: Bricks/util/test.cc (hunk 7 above)
Comment (dkorolev) [inline Bricks/util/test.cc]: You and your JavaScript style in C++. Still remember the suggestion to format ternary operator on three lines with an extra pair of parentheses :)

Comment (sompylasar) [inline Bricks/util/test.cc]: On five lines, actually :)
```
const x = (
  condition
    ? true
    : false
);
```
Can be shortened to four, but looks a little condensed.
```
const x = (condition
    ? true
    : false
);
```

Comment (dkorolev) [inline Bricks/util/test.cc]: Let's just agree an average C++ programmer knows a bit more about operator precedence compared to an average JavaScript one. The compiler reminding us to put semicolons where they belong kinda helps too. ;-)

Comment (sompylasar) [inline Bricks/util/test.cc]: Knowing more about operator precedence doesn't help much reading ternaries, especially nested :)

**Pull request merged**